packages/grafana-ui/src/types/datasource.ts

```diff
@@ -306,7 +326,25 @@ export interface DataStreamState {
   state: LoadingState;
 
   /**
-   * Consistent key across events.
+   * The key is used to identify unique sets of data within
+   * a response, and join or replace them before sending them to the panel.
+   *
+   * For example consider a query that streams four DataFrames (A,B,C,D)
+   * and multiple events with keys K1, and K2
+   *
+   * query(...) returns: {
+   *   state:Streaming
+   *   data:[A]
+   * }
+   *
+   * Events:
+   * 1. {key:K1, data:[B1]}    >> PanelData: [A,B1]
+   * 2. {key:K2, data:[C2,D2]} >> PanelData: [A,B1,C2,D2]
+   * 3. {key:K1, data:[B3]}    >> PanelData: [A,B3,C2,D2]
+   * 4. {key:K2, data:[C4]}    >> PanelData: [A,B3,C4]
+   *
+   * NOTE: that PanelData will report a `Done` state until all
+   * unique keys have returned with either `Error` or `Done` state.
```

Comment:
Whoa, are you saying that this is basically a stream of streams that you have to manually check by the key? Honestly did not see any code like that so far so not sure what to make from this. It seems like all of this is meant not to be used directly but only through `PanelQueryRunner` or something but that isn't the case right now so all these assumptions could be broken somewhere.

Reply:
> assumptions could be broken
well obviously!

> stream of streams
remember that the query argument is not a single query.  it is a request with a query[] (targets),  given that each query can return an array of data, that is how this mess gets made